Push image faster by using pigz to compress

lingoubb opened the pull request:
- closes https://github.com/moby/moby/issues/41987
- replaces / closes https://github.com/moby/moby/pull/42754
- related (containerd); https://github.com/containerd/containerd/issues/1896

The gzip compression used in the docker push has poor performance.
This change switches to trying to use pigz for gzip compression.
If it isn't available, it'll fall back to the default. This code path
can also be disabled by environment variable.

Signed-off-by: weileyuan <weileyuan@tencent.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
The gzip compression used in the docker push has poor performance. I changed it to using pigz for compression.
I noticed that docker used pigz to decompress the pull image for better performance (https://github.com/moby/moby/pull/35697), so I think the performance of docker push can be improved in the same way.

**- How I did it**
Similar to https://github.com/moby/moby/pull/35697, this change switches to trying to use pigz for gzip compression. If it isn't available, it'll fall back to the default. This code path can also be disabled by environment variable.

**- How to verify it**
Performance improvement verification: after the manual push test of some images (from 19.2 MB to 28.2 GB) , the time consumed by pushing images is reduced by 47.9%~74.7%. In fact, I have noticed that some issues have discussed and verified the performance improvement brought by this.
Correctness verification: There are existing tests for this codepath to ensure correctness of this implementation.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Push image faster by using pigz to compress

thaJeztah commented:
I know there's been prior discussion about using pigz for _push_, and I recall there were concerns about determinism (the potential of push with `pigz` producing a different artefact than without); so I'd like to get more eyes on this change.

/cc @tonistiigi @stevvooe @AkihiroSuda @cpuguy83

thaJeztah commented:
Looks like build is failing;

```
#48 0.186 Building: /build/bundles/cross/linux/arm/v5-daemon/dockerd
#48 0.186 GOOS="linux" GOARCH="arm" GOARM="5"
#48 33.74 # github.com/docker/docker/distribution
#48 33.74 distribution/push.go:105:15: undefined: os
#48 33.74 distribution/push.go:109:18: undefined: strconv
#48 33.74 distribution/push.go:117:20: undefined: exec
#48 33.74 distribution/push.go:154:9: undefined: exec
#48 33.74 distribution/push.go:159:13: undefined: bytes
```

thaJeztah commented on distribution/push.go:
```diff
@@ -126,3 +153,32 @@ func compress(in io.Reader) (io.ReadCloser, chan struct{}) {
+	}()
+
+	return pipeR, done
+}
```
Oh, while you're updating; could you also fix the missing newline here? 😅

lingoubb replied:
I'm very sorry. I will fix it in the next commit. (By the way, if I commit now, will the build restart?)

thaJeztah replied:
No worries! Yes, amending a commit, and force-pushing should restart the build. As you're a first-time contributor, I still need to manually approve GitHub Actions (but just did, so it should be running now)

lingoubb replied:
Thanks for the guidance, it looks like the checks have passed.

Commit:
Push image faster by using pigz to compress

The gzip compression used in the docker push has poor performance.
This change switches to trying to use pigz for gzip compression.
If it isn't available, it'll fall back to the default. This code path
can also be disabled by environment variable.

Signed-off-by: weileyuan <weileyuan@tencent.com>

samuelkarp commented:
> I know there's been prior discussion about using pigz for push, and I recall there were concerns about determinism

Do we know if pigz guarantees anything about the determinism for the resulting compressed file at specific compression settings?  Or how this relates to the existing output?

Because image layer sharing relies on a shared digest for the layer, and because the layer digest is computed based on the actual compressed bytes rather than the uncompressed bytes, any change to the method of compression can reduce the amount of effective sharing between images.  Additionally, with Docker's existing storage driver implementation discarding the original compressed layers (though this may change with the containerd backend migration; I haven't been following that as closely as I'd like), a change like this might increase the chances that you could pull some image to use as a base, build a new image, and push that new image (to a new registry/repository) without shared layers; in other words, the digests of the layers taken from your base image can change as the layer gets recompressed with a different compression implementation/settings.

lingoubb commented:
Do you mean that we need to ensure that pigz can get the same result every time when compressing the same file? Or does it also need to be consistent with gzip standard library?

lingoubb commented:
By the way, why not use uncompressed bytes to compute the digest

cpuguy83 commented:
> By the way, why not use uncompressed bytes to compute the digest

Because we aren't sending uncompressed bytes to the remote.


> Do you mean that we need to ensure that pigz can get the same result every time when compressing the same file? Or does it also need to be consistent with gzip standard library?

Both, pretty much. This is why we've rejected pigz for compression in the past, as I recall.
Once we get containerd's image store integrated we could possibly look at pigz (which is happening soon!). Once this is done docker won't be recompressing layers like it currently does and ultimately potentially messing up layer sharing.
With pigz it is all but a guarantee that we'll be breaking that sharing.

I didn't look at the code yet, but also would need to make it opt-in since pigz, as I understand it, will use significantly more cpu which  may not be desirable in some cases.

lingoubb commented:
This code path can be disabled by environment variable, or I can change it to off default.

Colossus commented:
Hey, thanks for working on this. I wanted to bump this, would be very interested in parallel compression

stevvooe commented on distribution/push.go:
```diff
@@ -100,7 +105,30 @@ func Push(ctx context.Context, ref reference.Named, config *ImagePushConfig) err
-func compress(in io.Reader) (io.ReadCloser, chan struct{}) {
+func compress(ctx context.Context, in io.Reader) (io.ReadCloser, chan struct{}) {
+	var usePigz bool
+	noPigzEnv := os.Getenv("MOBY_DISABLE_PIGZ")
```
I would recommend turning this off by default until we can prove it out.

stevvooe commented:
This has long been a suggestion. I'd say we'd need the following:

1. Ensure this is off by default.
2. Put in some kind of tests for determinism to ensure that multiple compression runs are deterministic.

This is probably ok for build scenarios where a daemon compresses and pushes once, so perhaps users should be able to make that decision on their own. However, the problems might be subtle to debug.

fellhorn commented:
> > I know there's been prior discussion about using pigz for push, and I recall there were concerns about determinism
> 
> Do we know if pigz guarantees anything about the determinism for the resulting compressed file at specific compression settings? Or how this relates to the existing output?
> 
> Because image layer sharing relies on a shared digest for the layer, and because the layer digest is computed based on the actual compressed bytes rather than the uncompressed bytes, any change to the method of compression can reduce the amount of effective sharing between images. Additionally, with Docker's existing storage driver implementation discarding the original compressed layers (though this may change with the containerd backend migration; I haven't been following that as closely as I'd like), a change like this might increase the chances that you could pull some image to use as a base, build a new image, and push that new image (to a new registry/repository) without shared layers; in other words, the digests of the layers taken from your base image can change as the layer gets recompressed with a different compression implementation/settings.

Mark Adler, the author of `pigz` answered this question before in https://github.com/madler/pigz/issues/90 .
For the same compression related options `pigz` produces deterministic results, which do not depend on the number of processes.

`pigz` does not produce the same results as `gzip` though. Based on @cpuguy83's comments I assume this means `pigz` compression can only be available behind a feature flag?

If done so is there much more blocking this PR?

thaJeztah commented:
closing as this is no longer in a mergable state (the `pkg/archive` was moved to a separate module), and this is less relevant with containerd integration, which pre-creates the compressed content
